### ISSUE TITLE
Hitting escape exits full screen

### ISSUE THREAD

**Issue by Hultner** [guyht/vimari#110]
As a person pretty much exclusively using safari in full/split screen mode I find myself constantly exiting fullscreen by accident when using vimari, for instance if I want to return to normal mode after writing a comment like this to hit the submit key from links hint mode.

This can be avoided by preventing default behaviour through an event listener.

Rough example:
```js
function disable_esc_exit(window){
    const ESC_KEY = 27;
    const KEYPRESS_EVENT = "keypress";

    window.addEventListener(KEYPRESS_EVENT, 
        event => event.keyCode === ESC_KEY && event.preventDefault()
    );
}

... 
disable_esc_exit(window);
```

**Comment by simeg** [comment]
Hi @Hultner. I don't understand your explanation of your use-case - could you provide a more detailed explanation, or maybe a gif? Thanks.

**Comment by Hultner** [comment]
Okay I'm going to list steps to reproduce 

1. Open Safari with vimari 
2. Enter Full Screen
3. Browse the web using vimari
4. Press <kbd>Esc</kbd> at some point

This trigger Safari to exit full screen mode, accompanied with a disruptive animation. 
In addition if the user ran the window in split Full Screen with another movie they need to rearrange the split windows manually again forcing the user to use their mouse or other pointer device.

**Comment by simeg** [comment]
Hi @Hultner, thanks for the detailed explanation! So if I understand correctly, the problem is that when you're using Safari in fullscreen and navigating with vimari and **accidentally hit escape one too many times**, fullscreen exits. Is that correct?

Or, is it that when you're in fullscreen and a **single esc key press** exits fullscreen? I can't really tell from your description.

Anyway, I made this GIF to demonstrate that the correct behavior is expected. I navigate Wikipedia with vimari and I can press esc to get out of link mode perfectly fine. But when I press esc when I'm not in some vimari mode - it exits fullscreen. I'm running version 1.13.

![vimari_safari_fullscreen](https://user-images.githubusercontent.com/8566054/44280780-fddf7480-a255-11e8-99db-c3e71262d9db.gif)

Let me know the exact problem and we'll take it from there.

**Comment by Hultner** [comment]
There’s two cases triggering this behaviour

1) I accidentally hit Esc to many times or due to muscle memory
2) I’ve edited a input field, hit Esc to exit insert mode to normal mode. My instinct when writing this comment is to input desired text, hit Esc, enter link target mode, trigger the [Comment]-button. 

Both these cases are equally disrupting. 

Thanks again for reading and taking my input into consideration.

**Comment by simeg** [comment]
Thanks @Hultner, I understand now. I can see that this is annoying and disruptive, and I think having an option for it makes sense. However, I'm torn between working on this current version of the extension (that will soon be deprecated due to the fact that it becomes incompatible with any Safari version below 11), and the new version of the extension which still lacks some basic stuff like settings.

I think implementing this would not be too difficult, and since you are the only one wanting this feature so far I recommend you do it yourself. I can be of help reviewing and testing but I won't write any code. I'm sorry, but I have to priorities my time.

**Comment by hauleth** [comment]
What is status on that? I am mostly looking on a way to exit "insert mode". The flows goes like:

1. Visit for example GitHub
2. Hit `i` to enter "pass-through" or "insert" mode (for example to use GitHub `y` shortcut to expand the URL)
3. Hit `<Esc>` to go back to "normal mode"

Workaround in some cases is to use `^[`, but on my keyboard (40% Planck) it is more troublesome than just using `<Esc>`.

**Comment by Hultner** [comment]
@hauleth I'm not working on it, I don't know if Vimari is actively being developed. At this moment I'm hoping the chrome based vim extensions will work on the next Safari version as they seem more mature and feature complete. I do love vimari, without it I wouldn't be able to use Safari as my daily browser but it's still missing a good chunk of the power from it's siblings in Chrome & Firefox.

**Comment by paddor** [comment]
What's the status here? This bug makes this extension pretty much useless in fullscreen mode.

**Comment by si3792** [comment]
I'm in the same boat — this basically forces me to choose between using the extension and full screen :(

**Comment by Markhenn** [comment]
Here is a little work around in case you use a tool like Better Touch Tool (BTT). There you can rebind the <kbd>esc</kbd>  key in Safari to something else. For me <kbd>esc</kbd> will now trigger `ctrl+[` which will tell vimari to go into normal mode instead of exiting full screen.

**Comment by ylorn** [comment]
is there an update on this? Hitting ESC should switch off the link mode instead of exiting the fullscreen

**Comment by 3nt3** [comment]
Hey, is there any update on this?

**Comment by 3nt3** [comment]
> is there an update on this? Hitting ESC should switch off the link mode instead of exiting the fullscreen

It does, it just also exits full screen if you do it one to many times.

**Comment by moritzreiter** [comment]
> > is there an update on this? Hitting ESC should switch off the link mode instead of exiting the fullscreen
> 
> It does, it just also exits full screen if you do it one to many times.

  Weirdly, for me, `esc` always exits full screen even if I don't hit it one too many times.

So, to spell it out:

1. I'm in Safari in full screen with an open web page
2. I press `f` to enter nav mode
3. I press `esc` _exactly __one__ time_ to exit nav mode
   - Boom: Safari exits full screen _and_ Vimari exits nav mode

So annoying 😭

**Comment by Hultner** [comment]
Sharing as I just recently set up a new computer. My current solution for this now 8 year old ongoing issue is the following karabiner rule (it remaps both caps lock press and escape in Safari as I've got caps-lock bound to esc via hyper, you can skip the caps lock rules if you don't). This rule effectively rebinds esc to ^[ in Safari only.

```json
{
    "description": "Safari: Escape/CapsLock → Ctrl+[ (vim-compatible, prevents fullscreen exit)",
    "manipulators": [
        {
            "type": "basic",
            "from": {
                "key_code": "escape",
                "modifiers": { "optional": ["any"] }
            },
            "to": [
                { "key_code": "open_bracket", "modifiers": ["left_control"] }
            ],
            "conditions": [
                {
                    "type": "frontmost_application_if",
                    "file_paths": ["/Applications/Safari.app/Contents/MacOS/Safari"]
                }
            ]
        },
        {
            "type": "basic",
            "from": {
                "key_code": "caps_lock",
                "modifiers": { "optional": ["any"] }
            },
            "to": [
                {
                    "key_code": "right_shift",
                    "modifiers": ["right_command", "right_control", "right_option"]
                }
            ],
            "to_if_alone": [
                { "key_code": "open_bracket", "modifiers": ["left_control"] }
            ],
            "conditions": [
                {
                    "type": "frontmost_application_if",
                    "file_paths": ["/Applications/Safari.app/Contents/MacOS/Safari"]
                }
            ]
        }
    ]
}
```
